cmd/kubeadm/app/cmd/join_test.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bytes"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmapiext "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+)
+
+const (
+	TestConfig = `apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority-data:
+    server: localhost:9008
+  name: prod
+contexts:
+- context:
+    cluster: prod
+    namespace: default
+    user: default-service-account
+  name: default
+current-context: default
+kind: Config
+preferences: {}
+users:
+- name: kubernetes-admin
+  user:
+    client-certificate-data:
+    client-key-data:
+`
+)
+
+var ValidNodeConfig = &kubeadm.NodeConfiguration{
+	Token:                      "abcdef.1234567890123456",
+	TLSBootstrapToken:          "abcdef.1234567890123456",
+	DiscoveryToken:             "abcdef.1234567890123456",
+	DiscoveryTokenAPIServers:   []string{"localhost:9008"},
+	DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+	CACertPath:                 "/some/cert.crt",
+}
+
+func TestValidateJoinCommandLine(t *testing.T) {
+	// create temp directory
+	tmpDir, err := ioutil.TempDir("", "kubeadm-join-test")
+	if err != nil {
+		t.Errorf("Unable to create temporary directory: %v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// create config file
+	configFilePath := filepath.Join(tmpDir, "test-config-file")
+	cfgFile, err := os.Create(configFilePath)
+	if err != nil {
+		t.Errorf("Unable to create file %q: %v", configFilePath, err)
+	}
+	defer cfgFile.Close()
+
+	testCases := []struct {
+		name                  string
+		args                  []string
+		skipPreFlight         bool
+		cfgPath               string
+		configToWrite         string
+		featureGatesString    string
+		ignorePreflightErrors []string
+		testJoinValidate      bool
+		testJoinRun           bool
+		cmdPersistentFlags    map[string]string
+		nodeConfig            *kubeadm.NodeConfiguration
+		expectedError         bool
+	}{
+		{
+			name:          "valid: join command line",
+			skipPreFlight: true,
+			cfgPath:       configFilePath,
+			configToWrite: TestConfig,
+			expectedError: false,
+		},
+		{
+			name:          "invalid: missing config file",
```

Comment:
Why missing config file is an invalid case? `--config` is not a mandatory flag. Am I missing something here?

Reply:
`missing-path-to-a-config` points to a missing file. the idea here is to enter this branch but fail the `ioutil.ReadFile()` call:
https://github.com/neolit123/kubernetes/blob/12b606d5ada2c4a09d5789387994002f2773cc9d/cmd/kubeadm/app/cmd/join.go#L213

Reply:
SGTM. I mis-read the test case :)